evalml/automl/automl_search.py

```diff
@@ -495,96 +517,111 @@ def search(self, data_checks="auto", show_iteration_plot=True):
             logger.info("Will stop searching for new pipelines after %d seconds.\n" % self.max_time)
         logger.info("Allowed model families: %s\n" % ", ".join([model.value for model in self.allowed_model_families]))
-        search_iteration_plot = None
+        self.search_iteration_plot = None
         if self.plot:
-            search_iteration_plot = self.plot.search_iteration_plot(interactive_plot=show_iteration_plot)
+            self.search_iteration_plot = self.plot.search_iteration_plot(interactive_plot=show_iteration_plot)
 
         self._start = time.time()
 
-        should_terminate = self._add_baseline_pipelines()
-        if should_terminate:
-            return
+        try:
+            self._add_baseline_pipelines()
+        except KeyboardInterrupt:
+            if self._handle_keyboard_interrupt():
+                self._interrupted = True
 
         current_batch_pipelines = []
         current_batch_pipeline_scores = []
-        while self._check_stopping_condition(self._start):
+        new_pipeline_ids = []
+        loop_interrupted = False
+        while self._should_continue():
             try:
-                if current_batch_pipeline_scores and np.isnan(np.array(current_batch_pipeline_scores, dtype=float)).all():
-                    raise AutoMLSearchException(f"All pipelines in the current AutoML batch produced a score of np.nan on the primary objective {self.objective}.")
-                current_batch_pipelines = self._automl_algorithm.next_batch()
-                current_batch_pipeline_scores = []
+                if not loop_interrupted:
+                    current_batch_pipelines = self._automl_algorithm.next_batch()
             except StopIteration:
                 logger.info('AutoML Algorithm out of recommendations, ending')
                 break
-
-            current_batch_size = len(current_batch_pipelines)
-            current_batch_pipeline_scores = self._evaluate_pipelines(current_batch_pipelines, search_iteration_plot=search_iteration_plot)
-
-            # Different size indicates early stopping
-            if len(current_batch_pipeline_scores) != current_batch_size:
-                break
-
+            try:
+                new_pipeline_ids = self._engine.evaluate_batch(current_batch_pipelines)
+                loop_interrupted = False
+            except KeyboardInterrupt:
+                loop_interrupted = True
+                if self._handle_keyboard_interrupt():
+                    break
+            full_rankings = self.full_rankings
+            current_batch_idx = full_rankings['id'].isin(new_pipeline_ids)
+            current_batch_pipeline_scores = full_rankings[current_batch_idx]['score']
+            if len(current_batch_pipeline_scores) and current_batch_pipeline_scores.isna().all():
+                raise AutoMLSearchException(f"All pipelines in the current AutoML batch produced a score of np.nan on the primary objective {self.objective}.")
```

Comment:
I moved the keyboard interrupt handling up to `AutoMLSearch.search`. I think `AutoMLSearch` should be solely responsible for handling keyboard interrupt, and the engines should be responsible for only adding completed pipelines to the results.

There's a bigger project hiding in here... if we really wanted atomicity of pipeline evaluations, we'd want to build a queue structure which can track when jobs error out and allow `AutoMLSearch` to either reschedule them or ignore them.

Reply:
@chukarsten that's a good point. At the moment I think its more of a nebulous long-term concern; I'd like to see how the parallel project shakes out before we file anything for it.